plugins/inventory/icinga2.py

```diff
@@ -0,0 +1,223 @@
+            data_dict['attrs'] = attrs
+        if joins is not None:
+            data_dict['joins'] = joins
+        if host_filter is not None:
+            data_dict['filter'] = host_filter
+        host_dict = self._post_request(query_hosts_url, data_dict)
+        return host_dict
+
+    def get_inventory_from_icinga(self):
+        """Query for all hosts """
+        self.display.vvv("Querying Icinga2 for inventory")
+        query_args = {
+            "attrs": ["address", "state_type", "state", "groups"],
+        }
+        # Icinga2 API Call
+        results_json = self._query_hosts(**query_args)['results']
+        ansible_inv = self._convert_inv(results_json)
+        return ansible_inv
+
+    def _populate(self):
+        groups = self._to_json(self.get_inventory_from_icinga())
+        return groups
+
+    def _to_json(self, in_dict):
+        """Convert dictionary to JSON"""
+        return json.dumps(in_dict, sort_keys=True, indent=2)
+
+    def _convert_inv(self, json_data):
+        """Convert Icinga2 API data to JSON format for Ansible"""
+        groups_dict = {"_meta": {"hostvars": {}}}
+        for entry in json_data:
+            host_name = entry['name']
+            host_attrs = entry['attrs']
+            if host_attrs['state'] == 0:
+                host_attrs['state'] = 'on'
+            else:
+                host_attrs['state'] = 'off'
+            host_groups = host_attrs['groups']
+            host_addr = host_attrs['address']
+            self.inventory.add_host(host_addr)
+            for group in host_groups:
+                if group not in self.inventory.groups.keys():
+                    self.inventory.add_group(group)
+                self.inventory.add_child(group, host_addr)
+            self.inventory.set_variable(host_addr, 'address', host_addr)
+            self.inventory.set_variable(host_addr, 'hostname', host_name)
+            self.inventory.set_variable(host_addr, 'state',
+                                        host_attrs['state'])
+            self.inventory.set_variable(host_addr, 'state_type',
+                                        host_attrs['state_type'])
+        return groups_dict
+
+    def parse(self, inventory, loader, path, cache=True):
+
+        super(InventoryModule, self).parse(inventory, loader, path)
+
+        # read config from file, this sets 'options'
+        self._read_config_data(path)
+        try:
+            # Store the options from the YAML file
+            self.icinga2_url = self.get_option('url').rstrip('/') + '/v1'
+            self.icinga2_user = self.get_option('user')
+            self.icinga2_password = self.get_option('password')
+            self.ssl_verify = self.get_option('validate_certs')
+            # Not currently enabled
+            # self.host_filter = self.get_option('host_filter')
+            # self.cache_key = self.get_cache_key(path)
+            # self.use_cache = cache and self.get_option('cache')
+        except Exception as error:
```

Comment:
What's the exception you are expecting here?

Reply:
Thinking about it, this doesn't seem necessary since the inventory plugin handles the missing options already.